### PR TITLE
2 duration fields in Testcase class

### PR DESCRIPTION
**Pull request by angelinag**
#1923

### CODE REVIEW

**Review by atodorov — CHANGES_REQUESTED** [review]
* squash all commits together, see http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html
* Add the existing issue number in commit log, see https://docs.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords
* Adjust commit log as necessary ^^^

File: tcms/testcases/migrations/0018_add_testcase_duration_fields.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# Generated by Django 3.1.6 on 2021-02-16 09:40
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('testcases', '0017_rename_related_names'),
```
Comment (atodorov) [inline tcms/testcases/migrations/0018_add_testcase_duration_fields.py:9]: See lint/black failures.  This code base uses `black` to reformat the code automatically. Type `black <filename>` on your machine and it should format the code for you.

**Review by atodorov — CHANGES_REQUESTED** [review]
Needs formatting changes.

Also reference the issue number inside the commit log, see previous comment.

File: tcms/testcases/templates/testcases/get.html [hunk 2]
```diff
@@ -83,6 +83,30 @@ <h2 class="card-pf-title" style="text-align: left">
                 <span class="fa fa-calendar"></span>{{ object.create_date }}
             </h2>
 
+            <h2 class="card-pf-title" style="text-align: left">
+                {% with dummy_setup_duration=3000 %}
```
Comment (atodorov) [inline tcms/testcases/templates/testcases/get.html:87]: Remove the `with` tag and display object properties directly. If you need real data you can update the fields directly in the DB:
```
./manage.py shell
from tcms.testcases.models import *

tc = TestCase.objects.get(pk=1234)
tc.setup_duration = timedelta(seconds=120)
tc.save()
```

File: tcms/testcases/templates/testcases/get.html [hunk 3]
```diff
@@ -83,6 +83,30 @@ <h2 class="card-pf-title" style="text-align: left">
...
+                <span class="fa fa-clock-o"></span>{% trans 'Setup duration' %}:
+                {{ dummy_setup_duration }}
+                {% endwith %}
+               <!-- TODO: filter for duration readability -->
```
Comment (atodorov) [inline tcms/testcases/templates/testcases/get.html:91]: Remove the TODOs - will make the tools happy and the list of items is also available in the issue description.

File: tcms/testcases/templates/testcases/mutable.html [hunk 4]
```diff
@@ -197,6 +209,7 @@
                 </div>
                 <div class="col-md-3 col-lg-3">
                     <input class="bootstrap-switch" name="{{ notify_formset.prefix }}-{{ forloop.counter0 }}-notify_on_case_update" type="checkbox" {% if notify_form.notify_on_case_update.value %}checked{% endif %}  data-on-text="{% trans 'Yes' %}" data-off-text="{% trans 'No' %}">
+                    <input class="bootstrap-switch" name="{{ notify_formset.prefix }}-{{ forloop.counter0 }}-notify_on_case_update" type="checkbox" {% if notify_form.notify_on_case_update.value %}checked{% endif %}  data-on-text="{% trans 'Yes' %}" data-off-text="{% trans 'No' %}">
```
Comment (atodorov) [inline tcms/testcases/templates/testcases/mutable.html:212]: Loos like a duplicate of the previous line, remove it.

**Review by atodorov — CHANGES_REQUESTED** [review]
Some changes requested. 

IMO focus on cleaning up the code review comments & displaying these fields in `testcases/get.html` first so that this PR can be merged. Some of the remaining items on the issue can prove to be bigger than anticipated and it is always better to work in smaller chunks.

Also see https://github.com/kiwitcms/Kiwi/pull/2253#pullrequestreview-591143684

**Comment by angelinag** [comment]
image
Current readability status

File: tcms/testcases/templates/testcases/get.html [hunk 5]
```diff
@@ -83,6 +83,21 @@ <h2 class="card-pf-title" style="text-align: left">
...
+            </h2>
+
+            <h2 class="card-pf-title" style="text-align: left">
+                <span class="fa fa-clock-o"></span>{% trans 'Testing duration' %}:
+                {{ object.testing_duration }}
+            </h2>
+
+            <h2 class="card-pf-title" style="text-align: left">
+                <span class="fa fa-clock-o"></span>{% trans 'Expected duration' %}:
+                {{ object.expected_duration }}
```
Comment (atodorov) [inline tcms/testcases/templates/testcases/get.html:98]: Opening the page without assigning any values to these new fields leads to:

`TypeError: unsupported operand type(s) for +: 'NoneType' and 'NoneType'`

**Review by atodorov — CHANGES_REQUESTED** [review]

**Comment by atodorov** [comment]
> image
> Current readability status in `get.html`. For me it looks good enough because we already have a lot of information displayed about each test case and additional words can only make it heavier to read for the end user but if you think we should also add `hours`/`seconds`/`minutes` let me know and I'll do it directly in the template.

Changing the display format is of less importance ATM, leave that item for last.

File: tcms/testcases/models.py [hunk 6]
```diff
@@ -57,6 +57,14 @@ class TestCase(TCMSActionModel):
     requirement = models.CharField(max_length=255, blank=True, null=True)
     notes = models.TextField(blank=True, null=True)
     text = models.TextField(blank=True)
+    setup_duration = models.DurationField(db_index=True, null=True, blank=True) or 0
+    testing_duration = models.DurationField(db_index=True, null=True, blank=True) or 0
+
+    @property
+    def expected_duration(self):
+        if not self.setup_duration and not self.testing_duration:
```
Comment (atodorov) [inline tcms/testcases/models.py:65]: Condition is `or` because any of the 2 fields may be `None`. Alternatively swap the order of the return statements if you want to use an `and` expression.

File: tcms/testcases/models.py [hunk 7]
```diff
@@ -57,6 +57,14 @@ class TestCase(TCMSActionModel):
...
+            return 0
```
Comment (atodorov) [inline tcms/testcases/models.py:66]: This should be `timedelta(0)` otherwise this property will return values with 2 different data types.

File: tcms/testcases/models.py [hunk 8]
```diff
@@ -57,6 +57,14 @@ class TestCase(TCMSActionModel):
...
+        return self.setup_duration + self.testing_duration
```
Comment (atodorov) [inline tcms/testcases/models.py:67]: This property needs tests in `tcms/testcases/tests/test_models.py`.

File: tcms/testcases/models.py (hunk 6 above)
Comment (atodorov) [reply tcms/testcases/models.py:65]: Furthermore if one of these fields is None but the other one is not we may assume that there's no setup for example and the expected duration will be equal to the value of the other field alone.

File: tcms/testcases/templates/testcases/mutable.html [hunk 9]
```diff
@@ -100,6 +100,18 @@
...
+                <label class="col-md-1 col-lg-1">{% trans "Setup duration" %}</label>
+                <div class="col-md-3 col-lg-3">
+                    <!-- https://github.com/koss-lebedev/bootstrap-duration-picker -->
+                </div>
+
+                <label class="col-md-1 col-lg-1">{% trans "Testing duration" %}</label>
+                <div class="col-md-3 col-lg-3">
+                    <!-- https://github.com/koss-lebedev/bootstrap-duration-picker -->
+                </div>
+            </div>
```
Comment (atodorov) [inline tcms/testcases/templates/testcases/mutable.html:113]: Remove these changes for now since they need more work. You can back them up locally to save yourself some typing for your next PR.

**Review by atodorov — CHANGES_REQUESTED** [review]
- Changes requested
- need tests for the new property
- squash all commits together, see https://github.com/kiwitcms/Kiwi/pull/2253#pullrequestreview-591143684

File: tcms/testcases/models.py [hunk 10]
```diff
@@ -58,8 +58,8 @@ class TestCase(TCMSActionModel):
     requirement = models.CharField(max_length=255, blank=True, null=True)
     notes = models.TextField(blank=True, null=True)
     text = models.TextField(blank=True)
-    setup_duration = models.DurationField(db_index=True, null=True, blank=True)
-    testing_duration = models.DurationField(db_index=True, null=True, blank=True)
+    setup_duration = models.DurationField(null=True, blank=True)
+    testing_duration = models.DurationField(null=True, blank=True)
```
Comment (atodorov) [inline tcms/testcases/models.py:62]: Why did you remove `db_index=True` ?

File: tcms/testcases/tests/test_models.py [hunk 11]
```diff
@@ -28,6 +29,43 @@ def test_create_testcase_with_cyrillic_works_issue_1770(self):
...
+        testing_duration = case_random_duration.testing_duration
+        expected = setup_duration + testing_duration
+
+        case_random_duration.save()
+
+        case_random_duration.refresh_from_db()
+        self.assertEqual(case_random_duration.expected_duration, expected)
+
+    def test_empty_setup_duration_is_calculated_correctly(self):
+        case_empty_setup_duration = TestCaseFactory(setup_duration = None)
```
Comment (atodorov) [inline tcms/testcases/tests/test_models.py:47]: name this variable `test_case` to make the code more obvious. The test method name serves to describe what behavior is being exercised so variable names do not need to carry the same information.

File: tcms/testcases/tests/test_models.py [hunk 12]
```diff
@@ -28,6 +29,43 @@ def test_create_testcase_with_cyrillic_works_issue_1770(self):
...
+        testing_duration = case_empty_setup_duration.testing_duration
```
Comment (atodorov) [inline tcms/testcases/tests/test_models.py:48]: hard-code both fields in the factory constructor to avoid these temporary variables

File: tcms/testcases/tests/test_models.py [hunk 13]
```diff
@@ -28,6 +29,43 @@ def test_create_testcase_with_cyrillic_works_issue_1770(self):
...
+        case_empty_setup_duration.save()
+
+        case_empty_setup_duration.refresh_from_db()
```
Comment (atodorov) [inline tcms/testcases/tests/test_models.py:51]: Why do we need `.save()` and `.refresh_from_db()` methods here ?

File: tcms/testcases/tests/test_models.py [hunk 14]
```diff
@@ -28,6 +29,43 @@ def test_create_testcase_with_cyrillic_works_issue_1770(self):
...
+        case_empty_testing_duration.refresh_from_db()
+        self.assertEqual(case_empty_testing_duration.expected_duration, setup_duration)
+
+    def test_empty_durations_are_calculated_correctly(self):
+        case_no_durations = TestCaseFactory(setup_duration = None, testing_duration = None)
+        case_no_durations.save()
+
+        case_no_durations.refresh_from_db()
+        self.assertEqual(case_no_durations.expected_duration, None)
+
```
Comment (atodorov) [inline tcms/testcases/tests/test_models.py:68]: All of these scenarios are essentially the same with different input data. This can be implemented with a single test method and parameterized with hard-coded values.  See: `test_test_execution_statuses_can_be_deleted_but_one_must_remain` for example.

File: tcms/testcases/tests/test_models.py [hunk 15]
```diff
@@ -28,6 +29,43 @@ def test_create_testcase_with_cyrillic_works_issue_1770(self):
...
+        case_empty_testing_duration.refresh_from_db()
+        self.assertEqual(case_empty_testing_duration.expected_duration, setup_duration)
+
+    def test_empty_durations_are_calculated_correctly(self):
+        case_no_durations = TestCaseFactory(setup_duration = None, testing_duration = None)
+        case_no_durations.save()
+
+        case_no_durations.refresh_from_db()
+        self.assertEqual(case_no_durations.expected_duration, None)
```
Comment (atodorov) [inline tcms/testcases/tests/test_models.py:67]: This scenario is wrong. The expected value when both fields are None is `timedelta(0)`, see https://github.com/kiwitcms/Kiwi/pull/2253#discussion_r588265419

**Review by atodorov — CHANGES_REQUESTED** [review]

**Comment by atodorov** [comment]
Closed in favor of https://github.com/kiwitcms/Kiwi/pull/2314 and https://github.com/kiwitcms/Kiwi/pull/2326
